### DISCUSSION TITLE
Flow generator

### DISCUSSION THREAD

**Discussion by Nir-edence** [PrefectHQ/prefect#5871] [Q&A] [answered]
Hi everyone,
I'm trying to create a function that generates flows. I want to pass in a list of functions and return a flow object as a linear DAG. 
My problems start when I'm trying to make tasks pass data from one to another. here is an example of what I'm trying to do: 

```
funcs = [first_func, second_func, third_func, fourth_func]

def flow_generator(funcs):
    tasks = []
    prev_task = None
    flow = Flow('Name')
    for func in funcs:
        tasks.append(Task(func)) # this line doesn't declare a Task properly as well
        if prev_task is None:
            prev_task = task
        else:
            flow.set_dependencies(prev_task, downstream_task=[task])
            flow.add_edge(upstream_task = prev_task, downstream_task = task) # I'm not sure what to do to make data pass properly 
            prev_task = task
    return flow

```

I hope it was clear enough. 
Thanks for the help!

**Comment by NodeJSmith** [comment, marked as answer]
Hey @Nir-edence - you're actually very close! You need to use the `key` kwarg in `add_edge` to assign the upstream task to a specific keyword argument in the downstream task. In the below argument the result keyword argument in the run method of MyTask will receive the result that upstream_task returns.

```python

class MyTask(Task):
   def run(self, result):
      self.process(result)

flow.add_edge(upstream_task = prev_task, downstream_task = task, key = result)
````

Here's the documentation for that argument.
![image](https://user-images.githubusercontent.com/8505845/171959991-071c9ca0-ca34-485b-b538-fa5afab2727c.png)
https://docs.prefect.io/api/latest/core/flow.html#flow-2

**Reply by Nir-edence** [reply]
Thanks @NodeJSmith, I understand what was my problem now, I also needed to declare Tasks as you showed. I had declared tasks in a different way which didn't really work. 
Really appreciate it!
